bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/WDChannels.java

```diff
@@ -44,4 +59,117 @@ public synchronized WorkerDataChannel getWorkerDataChannel(String name) {
+        if (strand.channelCount == receiveFields.length) {
+            BMap<BString, Object> map = strand.workerReceiveMap;
+            strand.workerReceiveMap = null;
+            strand.channelCount = 0;
+            strand.setState(State.RUNNABLE);
+            return map;
+        }
+        return null;
+    }
+
+    public Object receiveDataAlternateChannels(Strand strand, String[] channels) throws Throwable {
+        Object result = null;
+        boolean allChannelsClosed = true;
+        for (String channelName : channels) {
+            WorkerDataChannel channel = getWorkerDataChannel(channelName);
+            if (!channel.isClosed()) {
+                allChannelsClosed = false;
+                result = channel.tryTakeData(strand, true);
+                if (result != null) {
+                    result = handleNonNullResult(channels, result, channel);
+                }
+            } else {
+                if (channel.getState() == WorkerDataChannel.State.AUTO_CLOSED) {
+                    errors.add((ErrorValue) ErrorUtils.createNoMessageError(channelName));
+                }
+            }
+        }
+        return processResulAndError(strand, channels, result, allChannelsClosed);
+    }
+
+    private Object handleNonNullResult(String[] channels, Object result, WorkerDataChannel channel) {
+        Object resultValue = getResultValue(result);
+        if (resultValue instanceof ErrorValue errorValue) {
+            errors.add(errorValue);
+            channel.close();
+            result = null;
+        } else {
+            closeChannels(channels);
+        }
+        return result;
+    }
+
+    private static Object getResultValue(Object result) {
+        if (result instanceof WorkerDataChannel.WorkerResult workerResult) {
+            return workerResult.value;
+        }
+        return result;
+    }
+
+    private Object processResulAndError(Strand strand, String[] channels, Object result, boolean allChannelsClosed) {
+        if (result == null) {
+            if (errors.size() == channels.length) {
+                result = errors.get(errors.size() - 1);
+            } else if (!allChannelsClosed) {
+                strand.setState(BLOCK_AND_YIELD);
+            }
+        } else {
+            strand.setState(State.RUNNABLE);
+        }
+        return getResultValue(result);
+    }
+
+    private void closeChannels(String[] channels) {
+        for (String channelName : channels) {
+            WorkerDataChannel channel = getWorkerDataChannel(channelName);
+            channel.close();
+            channel.setRemovable();
+        }
+    }
+
+    public synchronized void removeCompletedChannels(Strand strand, String channelName) {
+        if (this.wDChannels != null) {
+            WorkerDataChannel channel = this.wDChannels.get(channelName);
+            if (channel != null && (channel.callCount == 2 || channel.isRemovable)) {
```

Comment:
why do we specifically check for call count 2?

Reply:
We keep an integer variable to check for the completion of the send and receive action (for both positive and negative completions) to verify that the message passing is completed. So that, we can remove that channel from the cache. I will slightly change the implementation to rely only on `callCount` and remove the `isRemovable` variable.